### DISCUSSION TITLE
Can I use pocketbase to add live chat to existing app?

### DISCUSSION THREAD

**Discussion by zupat** [pocketbase/pocketbase#1660] [Q&A] [answered]
I have app with go backend, svelte and flutter front end. I’m wondering if I could integrate pocketbase seamlessly into my stack. 

My idea:
Simple pb table with userid and msg body
User send message to my go backend
I validate it and add it to pocketbase through the api
All clients subscribed gets the message

Will this approach work? I feel like I’m overlooking something…

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure what the _"I validate it and add it to pocketbase through the api"_ step involves, but yes, in general it should work.

Some caveats to the realtime api:

- The realtime events are sent only on changes (create/update/delete).

- There is no buffering or CRDT-like handling, meaning that events are sent in the order they were received/processed and in high concurrent load this could lead to slightly out of order messages, so you may have to sort the events additionally by a timestamp field (eg. `created/updated`) client-side before showing it to the user.

- If you expect a lot of realtime connected clients at the same time you'll have to increase the max allowed file descriptors limit, otherwise you may get errors like "_Too many open files_" or similar. On linux this is usually done with `ulimit`. You can check your current limits with `ulimit -a`. To set a new limit, you can do something like `ulimit -n 5500` (it could be also defined in the systemd with `LimitNOFILE`). The limit usually should be the "_the expected concurrent connections_" (eg. 5000) + "_some reasonable threshold for the other regular connections, internal db and file operations_" (eg. ~500)

**Reply by dschissler** [reply]
Another way to scale chat is to create a single [Shared Worker](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker) that exists within the domain as long as a tab is open.  Then you have to use one of the forms of IPC.  So if someone opens a chat in two different windows then it's only using a single EventSource connection.  This also makes it possible to have notifications on other windows without creating another connection.

It would take a while to implement this stuff, so one way of moving forward could be to read all about it and then design the internal API in such a way that it could be swapped out later.  Like for example not connecting directly to the EventSource events, but having a middle layer there, even it if only just listens to an event and then sends it off again.
